torchserve/datadog_checks/torchserve/model_discovery.py

```diff
@@ -0,0 +1,67 @@
+# (C) Datadog, Inc. 2023-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from datadog_checks.base import AgentCheck
+from datadog_checks.base.checks.kubelet_base.base import urljoin
+from datadog_checks.base.utils.discovery import Discovery
+
+
+class ModelDiscovery(Discovery):
+    def __init__(self, check, limit=None, include=None, exclude=None, interval=None):
+        super().__init__(
+            self.get_models,
+            limit=limit,
+            include={pattern: None for pattern in include} if include else None,
+            exclude=exclude,
+            interval=interval,
+            key=lambda n: n.get("modelName"),
+        )
+        self.check = check
+        self.api_status = AgentCheck.UNKNOWN
+        self.all_models = None
+        # Used to know when the cache used internally in `Discovery` gets refreshed.
+        # Useful to avoid trying to parse the payloads to generate events in case it's not needed
+        self.refreshed = False
+
+    def get_items(self):
+        # Force the flag to False to make sure we do not keep the previous value.
+        self.refreshed = False
+        return super().get_items()
+
+    def get_models(self):
+        # The cache wanted to refresh, so we update the flag
+        self.refreshed = True
```

Comment:
Perhaps we can move `self.refreshed = True` to the end of the function in case there is an exception in updating the models in the middle.

Reply:
Sounds good to me!